lib/Analysis/NoiseAnalysis/BGV/Noise.cpp

```diff
@@ -0,0 +1,142 @@
+#include "lib/Analysis/NoiseAnalysis/BGV/Noise.h"
+
+#include <cmath>
+
+#include "lib/Analysis/NoiseAnalysis/Params.h"
+#include "llvm/include/llvm/Support/Debug.h"        // from @llvm-project
+#include "llvm/include/llvm/Support/raw_ostream.h"  // from @llvm-project
+
+#define DEBUG_TYPE "BGVNoise"
+
+namespace mlir {
+namespace heir {
+namespace bgv {
+
+template <bool W>
```

Comment:
Please add a code comment linking to the source for these formulas. Either a paper or code reference.